python/iceberg/parquet/parquet_reader.py

```diff
@@ -0,0 +1,264 @@
+     TypeID.DOUBLE: lambda field: (pa.float64(), np.nan),
+     TypeID.FIXED: lambda field: pa.binary(field.length),
+     TypeID.FLOAT: lambda field: (pa.float32(), np.nan),
+     TypeID.INTEGER: lambda field: (pa.int32(), np.nan),
+     TypeID.LIST: lambda field: (pa.list_(pa.field("element",
+                                                   DTYPE_MAP[field.type.element_type.type_id](field.type)[0])),
+                                 None),
+     TypeID.LONG: lambda field: (pa.int64(), np.nan),
+     # To-Do: update to support reading map fields
+     # TypeID.MAP: lambda field: (,),
+     TypeID.STRING: lambda field: (pa.string(), ""),
+     TypeID.STRUCT: lambda field: (pa.struct([(nested_field.name,
+                                               DTYPE_MAP[nested_field.type.type_id](nested_field.type)[0])
+                                              for nested_field in field.type.fields]), {}),
+     TypeID.TIMESTAMP: lambda field: (pa.timestamp("us"), datetime.now()),
+     # not used in SPARK, so not implementing for now
+     # TypeID.TIME: pa.time64(None)
+     }
+
+FS_MAP: typing.Dict[typing.Type[FileSystem], typing.Type[fs.FileSystem]] = {LocalFileSystem: fs.LocalFileSystem}
+
+try:
+    FS_MAP[S3FileSystem] = fs.S3FileSystem
+except ImportError:
+    _logger.warning("S3 Filesystem not available to arrow")
+
+
+class ParquetReader(object):
+
+    def __init__(self, input: InputFile, expected_schema: Schema, options, filter_expr: Expression,
+                 case_sensitive: bool, start: int = None, end: int = None):
+        self._stats: typing.Dict[str, int] = dict()
+
+        self._input = input
+        self._input_fo = input.new_fo()
+
+        self._arrow_file = pq.ParquetFile(self._input_fo)
+        self._file_schema = convert_parquet_to_iceberg(self._arrow_file)
+        self._expected_schema = expected_schema
+        self._file_to_expected_name_map = ParquetReader.get_field_map(self._file_schema,
+                                                                      self._expected_schema)
+        self._options = options
+        self._filter = get_dataset_filter(filter_expr, ParquetReader.get_reverse_field_map(self._file_schema,
+                                                                                           self._expected_schema))
+
+        self._case_sensitive = case_sensitive
+        if start is not None or end is not None:
+            raise NotImplementedError("Partial file reads are not yet supported")
+            # self.start = start
+            # self.end = end
+
+        self.materialized_table = False
+        self.curr_iterator = None
+        self._table = None
+        self._df = None
+        self._batches = None
+        self._row_tuple = None
+
+        _logger.debug("Reader initialized for %s" % self._input.path)
+
+    @property
+    def stats(self) -> dict:
+        return dict(self._stats)
+
+    def to_pandas(self) -> typing.Union[pd.Series, pd.DataFrame]:
```

Comment:
were you still going to remove the pandas and iterator stuff? I was kinda thinking that we would have a class 1 level higher that would read across all tables and pandas would be accessed there

Reply:
It's removed in the latest commit, and the to_arrow_table method has been renamed to `read()`.  Yes, there should be a something between `TableScan` and `ParquetReader` that combines the results of all the `ScanTask`s.